core/src/main/java/org/elasticsearch/rest/RestController.java

```diff
@@ -308,11 +325,115 @@ boolean checkRequestParameters(final RestRequest request, final RestChannel chan
         return true;
     }
 
-    private RestHandler getHandler(RestRequest request) {
+    void tryAllHandlers(final RestRequest request, final RestChannel channel, final ThreadContext threadContext) throws Exception {
+        for (String key : headersToCopy) {
+            String httpHeader = request.header(key);
+            if (httpHeader != null) {
+                threadContext.putHeader(key, httpHeader);
+            }
+        }
+        // Between retrieving the correct path, we need to reset the parameters,
+        // otherwise parameters are parsed out of the URI that aren't actually handled.
+        final Map<String, String> originalParams = new HashMap<>(request.params());
+
+        // Request execution flag
+        boolean requestHandled = false;
+
+        /*
+         * First try handlers mapped to explicit paths only.
+         */
+        if (requestHandled == false) {
+            final RestHandler handler = getHandler(request, TrieMatchingMode.EXPLICIT_NODES_ONLY);
```

Comment:
I experimented with doing this for quite a while the other day, unfortunately, it turned out to be too difficult to do cleanly with the way that `PathTrie` currently behaves with `RestHandler`. Right now attempting to retrieve a handler from the trie actually modifies the `RestRequest`, which means that trying to return a list of handlers is very side-effect-y and even if we select the right handler, there request may be out of sync because it was modified so many times.